src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs

```diff
@@ -31,63 +57,133 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
                 Name = "Console logger queue processing thread"
             };
             _outputThread.Start();
+            _meter.CreateObservableGauge<long>("queue-size", GetQueueSize);
         }
 
         public virtual void EnqueueMessage(LogMessageEntry message)
         {
-            if (!_messageQueue.IsAddingCompleted)
-            {
-                try
-                {
-                    _messageQueue.Add(message);
-                    return;
-                }
-                catch (InvalidOperationException) { }
-            }
-
-            // Adding is completed so just log the message
-            try
+            // cannot enqueue when adding is completed
+            if (!Enqueue(message))
             {
                 WriteMessage(message);
             }
-            catch (Exception) { }
         }
 
         // for testing
         internal void WriteMessage(LogMessageEntry entry)
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            try
+            {
+                var messagesDropped = Interlocked.Exchange(ref _messagesDropped, 0);
+                if (messagesDropped != 0)
+                {
+                    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}");
```

Comment:
>System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}"); [](http://example.com/codeflow?start=20&length=191)

Is it possible this message can spam the logs? I imagine showing this every time a message dropped can be a lot? is it right?

Reply:
It's at most every other message, right? If I'm dropping logs between every message, I'd want that to be very clear. I don't like emitting this message any less.

Reply:
it doesn't emit every time a message is dropped, but every other logs (could dropping 100, 900, etc. at a time)

Reply:
Exactly!

It's also tempting to write this error exactly where the N logs were dropped. I could see that being *very* useful for debugging.

I understand we're trying to avoid adding more load to an already-overloaded system. In API review we discussed not wanting to add entries to the queue which might effectively halve the queue length. Still, I think it might be better to keep a reference to the tail `LogMessageEntry` in the `ConsoleLogProcessor` and store the `_messagesDropped` in that. `Enqueue` we could do something like:

```csharp
while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
{
    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
    {
        Interlocked.Increment(ref _tailLogMessageEntry._messagesDropped);
        return true;
    }
```

And then here we could do something like:

```csharp
if (entry._messagesDropped != 0)
{
    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.");
}
```

I'm handwaving over how `_tailLogMessageEntry`, but this seems doable.